### DISCUSSION TITLE
Reticulum Application-Level Protocols

### DISCUSSION THREAD

**Discussion by 4c3e** [markqvist/Reticulum#12] [Ideas]
Hello all! As Reticulum development continues (a big thank you to @markqvist!) I wanted to open up a discussion about an internet protocol, and other application-level protocols that can be designed on top of Reticulum. LXMF and Nomadnet already seem to implement a sort of internet protocol that allows for a node's pages to be requested by clients. 

What I'm interested in is a [Gemini](https://gemini.circumlunar.space/docs/specification.gmi)-like protocol that creates connections using Reticulum's Link API. Gemini is enjoyable over the traditional internet, but it's simplicity and lightweight spec could actually be very useful on a high latency, low bandwidth systems, like Reticulum over packet radio, compared to traditional http and html. 

Another application-level protocol I'm interested in is a telnet/ssh virtual terminal tool that is also built over Reticulum Links. 

Currently I don't have much free time, but I'll be trying to work on a basic Reticulum Internet Protocol spec sheet here:  
https://github.com/4c3e/rip-spec

Edit: HTTP/3 is adding an [interesting streaming protocol](https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-22#section-1) that works over a UDP connection to multiplex different streams over a single connection. Certainly overkill for now, but a similar protocol could possibly work to enable multiple "Virtual Links" over a single underlying RNS Link.

**Comment by markqvist** [comment]
Thanks for starting this discussion @4c3e. I am not familiar with Gemini, but I just read your spec draft. A genereal purpose protocol like this would be very useful indeed. Since I am really busy right now, I am going to read it all again in more detail in a day or two, and also read up on Gemini and HTTP/3 to get a better basis for discussing this.

Am I correctly assuming that you intend to use the [Request API](https://markqvist.github.io/Reticulum/manual/reference.html#RNS.Link.request) as a carrier for the request/response flow?

Also, a remote shell utility is very high on my list of priorities :) It will probably be included as a bundled utility in one of the next releases :)

Link multiplexing is also a very interesting subject. For me the area of efficiently addressing multiple destinations over a single virtual link is especially interesting, since it will enable much better distributed communications systems. I'm working on the theory and methods for doing this with Reticulum, but it is probably a bit off still, since other things are more important to have working right now.

**Reply by 4c3e** [reply]
Been learning about Gemini for the past couple days. I think the part of it that's most useful here is the [Gemtext hyptertext specification](https://gemini.circumlunar.space/docs/gemtext.gmi). Basically it is an alternative the html that is extremely easy to parse, and is completely text-based. This allows users to easily write their own gemtext clients, and also keeps average page size (in bytes) very small, because all media files are always simply text links instead of automatically downloaded and displayed. None of the gemtext spec relies on the underlying Gemini protocol. I think it would make sense to have gemtext (or a slightly extended version of it that allows text input forms, which I have been working on) be the "default" hypertext mode for a general-use Reticulum internet. 

>Am I correctly assuming that you intend to use the Request API as a carrier for the request/response flow?

Thank you for pointing this out, not sure how I missed this in the API documentation. Using the request API makes more sense than trying to implement a custom Gemini protocol over a Reticulum Link, plus the request API is already implemented which is always helpful! Going forward I think the adding a similar post() function would allow for a more interactive system. However, for now I think the current Link, Resource, and Request API's are more than enough to implement a basic browser and server system on Reticulum. Going to be working on that here: https://github.com/4c3e/rip-demo-1

Edit: I noticed the "data" field in the request() api, could this be used to send a msgpack object, or some sort of binary blob?

Great news on the remote shell utility! And as for multiplexing: it's definitely out-of-scope for the near future but it's an interesting concept that seems to be useful. SSH also has a similar "channel" system that maintains a number of virtual connections all over a single TCP connection. So it seems "multiplexing" could be achieved at the application level right now, by sending and received labeled packets over a Link.

**Reply by markqvist** [reply]
It's interesting, in many ways, it looks like Gemini is trying to achieve many of the same things that I am with the _micron_ markup pages and text browser in built into Nomad Network. Unfortunately their site seems to be down right now, so haven't been able to read the spec. Maybe I would have used Gemini for Nomad Network instead of writing my own markup language if I had known about it ;) But micron _is_ very efficient though, so looking forward to reading the Gemini spec and seeing how they compare :)

Yes, the "data" field is for general purpose data posted as part of the request. It can be any size, and any type, a dictionary for example. Reticulum will automatically pack it in the most efficient way. It can also be plain text or binary data if you prefer, but for most purposes, it will probably be easiest to just use a list or a dictionary, since it will be auto-packed to the most efficient form.

**Reply by markqvist** [reply]
But really cool with the RIP server demo. Looking forward to seeing what this develops into! A general purpose page serving system is very useful indeed!

**Reply by 4c3e** [reply]
Here's a mirror for the gemtext spec: https://gitlab.com/gemini-specification/gemini-text/-/blob/master/specification.gmi

Micron looks very interesting, is it utf-8 based? And can you write micron documents using normal text editors? I have been interested in extending the Gemini spec to allow for more formatting options, which seems closer to what micron is than gemini. Specifically I've been looking into adding forms and inline links on top of the Gemini spec.

Also, the [browser/server minimal demo](https://github.com/4c3e/rip-demo-1) is working if anyone wants to check it out, note that the index.gem file has a hard-coded link that will need to be changed to the server's destination hash.

Edit: Just checked out the micron section of nomadnet's guides, really cool!

**Reply by 4c3e** [reply]
Also, when writing request handlers for RNS Links, what is the recommended way to return static files? Currently I am just reading text files into one string variable, then returning that string variable, which runs into issues with serving non-text files.

**Reply by 4c3e** [reply]
After messing around with the LXMF page system on Nomadnet, I think the best direction to go is to simply "open up" the protocols that nomadnet is using under the hood, instead of building a whole separate system. By "open up" I basically mean documenting what transactions happen under the hood for Nomadnet to request and display a page, so alternative stand-alone browsers and servers could be built and all talk to each other.

A side note, I think defining a standard message format would be useful as well, to allow for different messaging applications to communicate with each other as well. I'm pretty sure nomadnet conversations just use "raw" LXMF packets so I think there isn't much to this(?)

**Reply by 4c3e** [reply]
@markqvist et al, I updated the [RIP spec sheet](https://github.com/4c3e/rip-spec) to take into account the Request/Response API and Micron, would love some comments if anyone has any. The main differences between RIP and how pages are handled in Nomadnet (to my knowledge) is:
- URL (`rip://<destination>/path/goes/here` vs `<destination>:/path/goes/here`)
- RIP Response packets are msgpacked `[header, body]`. The header value allows clients to get a more granular response status and MIME type of the response body, and leaves extra room for custom protocols to be built using it. 

I built a small class to make synchronous requests in this method: https://github.com/4c3e/riprequest

**Reply by markqvist** [reply]
This is starting to look like something :) Just taking a day off today, but will return to this shortly. I have a lot of comments, thoughts and ideas, and I think we can shape this into something very useful. Thanks for bringing up this discussion.

**Comment by markqvist** [comment]
I regret that I have not yet gotten back to you on this. I have been exceedingly busy with work on other parts of the project. This will be short and not go into full details as I would have liked, so it is just a few comments.

I have been known to be quite pedantic at times, so don't take this as critique of the idea or spec itself. It is aimed at the wording and how the naming communicates purpose, which I think is extremely important.

I think the naming of this specification could potentially be misleading. The closest analogue to this spec in the "world of established protocols" would be HTTP, *HyperText Transfer Protocol*. The proposed spec seeks to define a structure for page/text transfer over Reticulum, providing the basic mechanism for what could be analogous to the *world wide web*. Yet it is named an *internet* protocol. I find this inaccurate.

Internet != world wide web != page transfer protocols. One could say that Reticulum itself, at its base, is an alternative *inter-net* protocol. It can be used to build both isolated, separate network, and connect these networks together arbitrarily. It allows both *networking* and *inter*-networking.

To illustrate what this spec is really seeking to provide, I think it is essential to update the naming to reflect it.

**Reply by 4c3e** [reply]
Great points, I definitely agree. I'll think on the naming, and if you or anyone else has any suggestions feel free to drop a reply!

**Reply by 4c3e** [reply]
"Reticulum Resource Transfer Protocol" might be a better fit here.

**Reply by markqvist** [reply]
I really like that :) It's concise and descriptive.

**Reply by 4c3e** [reply]
Great! Planning to jump back into some work on "RRTP" soon. 

Some quick thoughts I had over the past few weeks:
- Now I am leaning towards making the request and response system follow a strict subset of HTTP instead of something "third-party" like Gemini. This should make interoperability with the "outside internet" easier. HTTP request were also built with request bodies in mind, something Gemini does not support. A "POST request" system will give a nice framework for file uploading, centralized forums (alternative to distributed BBS's built on the LXMF propagation node system), and form submission etc. 
- Not sure if we want to support automatic destination redirects at the protocol level. Need to read more about how HTTP handles this.

**Reply by markqvist** [reply]
Sounds good!
- Interesting approach, if you make it a HTTP subset. It would make it a little bit easier to proxy/tunnel to the HTTP world, as you say, and would possibly come with other benefits as well. It might also impose annoying limitations that would prevent us from doing stuff we could have wanted further down the road. It definitely warrants some exploration, to figure those things out.
- I think we should be really careful about *automatic* redirects. In my opinion it's a can of worms that have spawned all kinds of workaround and strange situations in the HTTP world. Is it really a necessity? I think that no matter what, the user should always know that they are being diverted from the endpoint/resource they actually requested to someplace that is potentially completely different. Automatic redirects can be used to fool people in so many ways, and could lure unsuspecting users to give up anonymity to malicious actors.

**Reply by 4c3e** [reply]
@markqvist Sorry, I thought I wrote a response here, but I guess I didn't post it. 

I've went back and forth between my previous "HTTP subset" idea. Reading through HTTP header types, nothing jumps out at me as particularly useful to bake in at a protocol-level. Maybe the best move here is to make the standardized protocol as simple as possible, and move any kinds of advanced header-type information to the request/response body section?

For example, lets take the following request header format `TYPE URL META\n` where the meta field is optional
the two main formats for RRTP requests would look like:
`GET <89879879879879>/some/path\n`
`POST <89879879879879>/some/path application/msgpack\nREQUESTBODY` 
maybe could even switch GET and POST strings to just 0 and 1 to save some characters

Add in the same response header system as listed in the current RRTP document, and it's pretty much the minimal information you need for a system like this, and advanced features like cache timing, are pushed to the response/request body, like how the caching system currently works in micron. 

Also, as authentication is already implemented into Nomadnet, it might be worth it to add a standardized auth feature to the RRTP spec? I already added a few sentences on auth because of the identify() method in the Link API.

Edit: also agreed on automatic redirects

**Reply by markqvist** [reply]
Yes, I think I am a fan of that approach, keeping it as simple as possible, and letting any extra functionality that extends the "default use case" be specified by optional headers or extra fields. One option is to just include it in the response body, as you suggest. Or would it be better to define the response as one of two possibilities:

1. Just the response data. In this case it is interpreted according to a "default" case, for example "This is UTF-8 encoded text data for display, with no additional settings/headers".
2. An array containing two elements, the first being the response data. The second a dict containing headers/settings/some-other-name-for-it. One way to think of it could be an "extension" dict. That would make it very flexible and possible to implement all kinds of things. Maybe too open ended.

I don't really have any clear favorites or good arguments for what the best approach is here, so sorry for just throwing various ideas onto the pile ;)

**Reply by 4c3e** [reply]
I think borrowing Gemini's response header format is a good compromise of simplicity and functionality. For reference they use a single line of UTF-8 text, that takes the form `STATUS META\n` where STATUS is a status code, and META is a string depending on the status, with a newline char at the end. After the newline char, the response body start (if applicable). 

This adds very little overhead, but bakes in some very useful information to the protocol layer. The meta tag is used for error messages, and on success it is used for communicating the MIME type of the response body which can be used by receiving instance to correctly parse the Resource. 

If this sounds good, I can clean up the RRTP document, which we can then review as a whole, and then start working on a PR for tweaking nomadnet to use it :+1:

**Reply by markqvist** [reply]
I like the simplicity of it, but I also get the feeling that it is a bit strange to add the headers a text line to the response body. What if the response body is not text? What if it is an ELF binary, or JPG image? Or what if the response is itself encrypted with a key external to reticulum. In such cases, parsing out the header line and reconstructing the binary blob of response data seems a bit cumbersome. Maybe it would be better to have the header information in a variable separate from the main response body? The request/response API already supports this, since the response can be any data type, including a dict or an array, which is then automatically reconstructed as such on the receiving end before reaching the receiving program.

**Reply by 4c3e** [reply]
> What if the response body is not text?

When I first experimented with response headers I used an array that was then packed with msgpack, because I didn't want to be bothered with parsing text :laughing: 

If we use msgpack to encode a dict/array would an eventual C implementation of RNS be able to correctly unpack the data as well?

**Reply by markqvist** [reply]
Yes, msgpack has C implementations as well. But you don't need to use msgpack in your own code! Reticulum uses msgpack internally. You can just return a dict or array from the response callback, and Reticulum will transparently compress, pack, unpack and decompress it in the most efficient way.

**Reply by 4c3e** [reply]
Oh awesome, I guess one other question I have is if you can return a python file object in the response callback and it will also just work?

For example, lets say we handle response's by using the form `[header, body]` which is a python array that will get packed automatically like you mentioned. Could the `body` var be a python file object? For example's sake, lets say the response body is a `.png` image, and the header is the string: `20 image/png`

**Reply by markqvist** [reply]
As far as I remember, you can. Sorry, I'm travelling right now, so can't easily cgeck the source. That behaviour is the intention anyway, and I am pretty sure I did get around to implementing it. The request/response mechanism uses the Resource API internally, which *does* support passing file objects directly.

No matter what, you can count on this as the intended behaviour. If it's not working, I will fix it pronto :)

Edit: come to think of it, I am pretty sure it works when passing files directly as the response, but it *might* not work as members of an array. We should probably have that case working also though.

**Comment by 4c3e** [comment]
I have been thinking on this for a bit more, and I came up with another route which might make more sense. Would love other's input.

Similar to the optional title field in LXMF, it might be convenient to just specify a "META" field in the request API. For most RNS requests and responses, it would be useful to have access to some sort of metadata about the attached data.

It would slightly bloat situations where this metadata is not needed, but applications could just ignore the field in these cases.

**Reply by markqvist** [reply]
This is actually already possible, and something that was the intention from the beginning. Just in a bit of a different way than you might be used to. Consider the Request API:

```python
request(path, data=None, response_callback=None, failed_callback=None, progress_callback=None, timeout=None)
```

That `data` argument is not just for passing a string, or a blob of bytes, or something like that. Consider this:

```python
...

filename = "the_file.txt"
file = open(filename, "rb")

request_data = {
  "command": "upload",
  "data": file.read(),
  "meta": {
    "filename": filename,
    "creator": "somebody",
    "expiration_date": None,
  }
}

...
```

You can just pass that `request_data` dict directly to the Request API. Reticulum will handle packing, compressing, sequencing and mapping the entire thing in the most efficient way possible, completely automatically. In the other end the reverse happens, and the receiver just gets the exact same kind of dict passed to its callback.

So you don't really need to add any meta fields. It's already there. It's infinitely expandable, and you don't have to do anything extra. Just define the structures you need and send them :)

**Reply by 4c3e** [reply]
Apologies, I was having a bit of trouble understanding this bit of the request api, seeing the example you attached cleared things up. You're right, this already looks ideal!

**Reply by markqvist** [reply]
No worries, it absolutely bears elaborating on a bit, since it isn't at all obvious that Reticulum allows you to do this :)
